src/main/java/uk/nhs/hee/tis/trainee/forms/service/FormRelocateService.java

```diff
@@ -0,0 +1,182 @@
+import uk.nhs.hee.tis.trainee.forms.service.exception.ApplicationException;
+
+
+@Slf4j
+@Service
+@Transactional
+@XRayEnabled
+public class FormRelocateService {
+
+  private final FormRPartARepository formRPartARepository;
+  private final FormRPartBRepository formRPartBRepository;
+  protected final AmazonS3 amazonS3;
+
+  @Value("${application.file-store.bucket}")
+  private String bucketName;
+
+  /**
+   * Constructor for Form Relocate service.
+   *
+   * @param formRPartARepository  spring data repository (Form R Part A)
+   * @param formRPartBRepository  spring data repository (Form R Part B)
+   * @param amazonS3              client for S3 service
+   */
+  public FormRelocateService(FormRPartARepository formRPartARepository,
+                             FormRPartBRepository formRPartBRepository,
+                             AmazonS3 amazonS3,
+                             String bucketName) {
+    this.formRPartARepository = formRPartARepository;
+    this.formRPartBRepository = formRPartBRepository;
+    this.amazonS3 = amazonS3;
+    this.bucketName = bucketName;
+  }
+
+  /**
+   * Relocate Form R.
+   */
+  public void relocateFormR(String formId, String targetTrainee) throws IOException {
+
+    // Get Form R from MongoDB by FormId
+    AbstractForm formR = getMoveFormInfoInDb(formId);
+
+    if (formR == null) {
+      log.error("Cannot find FormR with ID " + formId + " from DB.");
+      throw new ApplicationException("Cannot find FormR with ID " + formId + " from DB.");
+    }
+    else {
+      String formType = formR.getFormType();
+      String sourceTrainee = formR.getTraineeTisId();
+      try {
+        updateTargetTraineeInDb(formR, targetTrainee);
+        if (formR.getLifecycleState() != LifecycleState.DRAFT) {
+          relocateFormInS3(formType, formId, sourceTrainee, targetTrainee);
+        }
+      } catch (Exception e) {
+        log.error("Fail to relocate FormR to target trainee: " + e + ". Rolling back...");
+        try {
+          updateTargetTraineeInDb(formR, sourceTrainee);
+          if (formR.getLifecycleState() != LifecycleState.DRAFT) {
+            relocateFormInS3(formType, formId, targetTrainee, sourceTrainee);
+          }
+        } catch (Exception ex) {
+          log.error("Fail to roll back: " + ex);
+        }
+        throw new ApplicationException("Fail to relocate FormR to target trainee: " + e.toString());
+      }
+    }
+  }
+
+  private AbstractForm getMoveFormInfoInDb(String formId) {
+    try {
+      Optional<FormRPartA> optionalFormRPartA =
+          formRPartARepository.findById(UUID.fromString(formId));
+      Optional<FormRPartB> optionalFormRPartB =
+          formRPartBRepository.findById(UUID.fromString(formId));
+
+      if (optionalFormRPartA.isPresent()) {
+        return optionalFormRPartA.get();
+      }
+      else if (optionalFormRPartB.isPresent()) {
+        return optionalFormRPartB.get();
+      }
+      return null;
```

Comment:
I'd prefer this method to return `Optional<AbstractForm>` instead of `null`

Reply:
>     try {
>       Optional<FormRPartA> optionalFormRPartA =
>           formRPartARepository.findById(UUID.fromString(formId));
>       if (optionalFormRPartA.isPresent()) {
>         return optionalFormRPartA.get();
>       }
>       else {
>         Optional<FormRPartB> optionalFormRPartB =
>             formRPartBRepository.findById(UUID.fromString(formId));
>         return optionalFormRPartB.get();
>       }
>     } catch (Exception e) {
>       log.error("Fail to get form with ID " + formId + ": " + e);
>       throw new ApplicationException("Fail to get form with ID " + formId + ": " + e.toString());
>     }

I was trying to return `Optional<AbstractForm>` but having issuing converting `Optional<FormRPartA>` into `Optional<AbstractForm>`. 
After refactoring, it is now returning AbstractForm with value optionalFormRPartA.get() or optionalFormRPartB.get(). When it is not found in both FormR Part A & B, it would intentionally throw exception. Please advise if there are better way to refactor it, please.

Reply:
Thank @Judge40. It is now updated to return `Optional<AbstractForm>` and handle if the option abstractForm is not present.